pkg/file/joiner/internal/job.go

```diff
@@ -0,0 +1,211 @@
+	doneC      chan struct{} // channel to signal termination of join loop
+	err        error         // read by the main thread to capture error state of the job
+	logger     logging.Logger
+}
+
+// NewSimpleJoinerJob creates a new simpleJoinerJob.
+func NewSimpleJoinerJob(ctx context.Context, store storage.Storer, rootChunk swarm.Chunk) *SimpleJoinerJob {
+	spanLength := binary.LittleEndian.Uint64(rootChunk.Data()[:8])
+	levelCount := getLevelsFromLength(int64(spanLength), swarm.SectionSize, swarm.Branches)
+
+	j := &SimpleJoinerJob{
+		ctx:        ctx,
+		store:      store,
+		spanLength: int64(spanLength),
+		dataC:      make(chan []byte),
+		doneC:      make(chan struct{}, 1),
+		logger:     logging.New(os.Stderr, 6),
+	}
+
+	// startLevelIndex is the root chunk level
+	// data level has index 0
+	startLevelIndex := levelCount - 1
+	j.data[startLevelIndex] = rootChunk.Data()[8:]
+	j.logger.Tracef("simple joiner start index %d for address %x", startLevelIndex, rootChunk.Address())
+
+	// retrieval must be asynchronous to the io.Reader()
+	go func() {
+		err := j.start(startLevelIndex)
+		if err != nil {
+			// this will only already be closed if all the chunk data has been fully read
+			// in this case the error will always be nil and this will not be executed
+			if err != io.EOF {
+				j.logger.Errorf("chunk join job fail: %v", err)
+			} else {
+				j.logger.Tracef("chunk join job eof")
+			}
+		}
+		j.err = err
+		close(j.dataC)
+		close(j.doneC)
+	}()
+
+	return j
+}
+
+// start processes all chunk references of the root chunk that already has been retrieved.
+func (j *SimpleJoinerJob) start(level int) error {
+
+	// consume the reference at the current cursor position of the chunk level data
+	// and start recursive retrieval down to the underlying data chunks
+	for j.cursors[level] < len(j.data[level]) {
+		err := j.nextReference(level)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// nextReference gets the next chunk reference at the cursor of the chunk currently loaded
+// for the specified level.
+func (j *SimpleJoinerJob) nextReference(level int) error {
+	data := j.data[level]
+	cursor := j.cursors[level]
+	chunkAddress := swarm.NewAddress(data[cursor : cursor+swarm.SectionSize])
+	err := j.nextChunk(level-1, chunkAddress)
+	if err != nil {
+		return err
+	}
+
+	// move the cursor to the next reference
+	j.cursors[level] += swarm.SectionSize
+	return nil
+}
+
+// nextChunk retrieves data chunks by resolving references in intermediate chunks.
+// The method will be called recursively via the nextReference method when
+// the current chunk is an intermediate chunk.
+// When a data chunk is found it is passed on the dataC channel to be consumed by the
+// io.Reader consumer.
+func (j *SimpleJoinerJob) nextChunk(level int, address swarm.Address) error {
+
+	// attempt to retrieve the chunk
+	j.logger.Tracef("next chunk level %d get: %v", level, address)
```

Comment:
Maybe to add more information about where this log message comes from, start with something like `simple joiner job: ...`. This goes for all log messages in this PR.